### PR TITLE
Add Naive Params() and ParamsAt()

### PR DESCRIPTION
**Pull request by blukat29**
## Proposed changes

- Add `MixedEngine.Params()` that returns the params out of `Governance.currentSet`.
  - This can replace `GetGovernanceValue(key)` and all nominal getters.
- Add `MixedEngine.ParamsAt(0)` that returns the params out of `db.ReadGovernanceAtNumber()`
  - This can replace `GetGovernanceItemAtNumber(num, key)`.

Intended use:

```diff
  // governance/api.go
- if GovernanceModeMap[api.governance.GovernanceMode()] == params.GovernanceMode_Ballot {
-     return true
- }
- return false
+ return api.governance.Params().GovernanceModeInt() == params.GovernanceMode_Ballot
```

```diff
  // consensus/istanbul/backend/snapshot.go
- if r, err := gov.GetGovernanceItemAtNumber(number, governance.GovernanceKeyMapReverse[params.Epoch]); err == nil && r != nil {
-     epoch = r.(uint64)
- } else {
-     logger.Error("Couldn't get governance value istanbul.epoch", "err", err, "received", r)
-     epoch = params.DefaultEpoch
- }
+ pset, err := gov.ParamsAt(number)
+ if err != nil {
+     logger.Error("Couldn't get governance value", "err", err)
+ }
+ epoch := pset.Epoch()
```

Actual replacement will happen in another PR, as it involves a lot of changes.

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

https://github.com/klaytn/klaytn/issues/1375

### CODE REVIEW

File: consensus/istanbul/backend/engine_test.go [hunk 1]
```diff
@@ -1327,3 +1336,95 @@ func TestGovernance_Votes(t *testing.T) {
...
+		chain, engine := newBlockChain(4, configItems...)
+
+		oldStakingManager := reward.GetStakingManager()
+		stakingInfo := makeFakeStakingInfo(0, nodeKeys, stakes)
+		reward.SetTestStakingManagerWithStakingInfoCache(stakingInfo)
+
+		var previousBlock, currentBlock *types.Block = nil, chain.Genesis()
+
+		// Create blocks with votes
+		for num := 0; num <= tc.length; num++ {
```
Comment (ian0371) [inline consensus/istanbul/backend/engine_test.go:1391]: Do we need to run the loop when num=9?

File: consensus/istanbul/backend/engine_test.go [hunk 2]
```diff
@@ -1327,3 +1336,95 @@ func TestGovernance_Votes(t *testing.T) {
...
+			assert.NoError(t, err)
+
+			// Load parameters for the next block
+			err = engine.governance.UpdateParams()
+			assert.NoError(t, err)
+		}
+
+		// Validate historic parameters with ParamsAt() and ReadGovernance().
+		// Check that both returns the expected result.
+		for num := 0; num <= tc.length; num++ {
```
Comment (ian0371) [inline consensus/istanbul/backend/engine_test.go:1417]: Do we need to run the loop when num=9?

File: consensus/istanbul/backend/engine_test.go [hunk 3]
```diff
@@ -668,6 +668,15 @@ func makeExpectedResult(indices []int, candidate []common.Address) []common.Addr
 	return copyAndSortAddrs(expected)
 }
 
+// Asserts taht if all (key,value) pairs of `subset` exists in `set`
+func assertMapSubset(t *testing.T, subset, set map[string]interface{}) {
+	if subset != nil {
```
[the remark this answers is not in the document]
Comment (blukat29) [reply consensus/istanbul/backend/engine_test.go]: Fixed in [97db580](https://github.com/klaytn/klaytn/pull/1380/commits/97db5809088b0653a0b346c69fcb896de60642e8).

File: consensus/istanbul/backend/engine_test.go (hunk 1 above)
Comment (blukat29) [reply consensus/istanbul/backend/engine_test.go:1389]: Fixed in [97db580](https://github.com/klaytn/klaytn/pull/1380/commits/97db5809088b0653a0b346c69fcb896de60642e8).

File: consensus/istanbul/backend/engine_test.go (hunk 2 above)
Comment (blukat29) [reply consensus/istanbul/backend/engine_test.go:1415]: Fixed in [97db580](https://github.com/klaytn/klaytn/pull/1380/commits/97db5809088b0653a0b346c69fcb896de60642e8).

**Review by ian0371 — APPROVED** [review]

File: governance/interface.go [hunk 4]
```diff
@@ -26,6 +26,25 @@ import (
 
 type Engine interface {
 	HeaderEngine
+	ReaderEngine
+}
+
+type ReaderEngine interface {
+	// Returns the current params. Which are the values that shall be
```
Comment (yoomee1313) [inline governance/interface.go:33]:
```suggestion
	// Returns the params at the current block. The returned params shall be
```

File: governance/interface.go [hunk 5]
```diff
@@ -26,6 +26,25 @@ import (
...
+	// should be using this method.
+	Params() *params.GovParamSet
+
+	// Returns the params at given block number. The returned params
+	// were used to build the block at given number.
+	// The number must be equal or less than current block height (head).
+	ParamsAt(num uint64) (*params.GovParamSet, error)
+
+	// Update the current params (the ones returned by Params()).
+	// by reaading the latest blockchain states.
```
Comment (yoomee1313) [inline governance/interface.go:44]:
```suggestion
	// by reading the latest blockchain states.
```

File: governance/mixed_test.go [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+	valueA := uint64(0x11)
+	valueB := uint64(0x22)
+
+	config := getTestConfig()
+	config.Istanbul.SubGroupSize = valueA
+	e, _, defaultGov := newTestMixedEngine(t, config)
+
+	// Write to database. Note that we must use gov.WriteGovernance(), not db.WriteGovernance().
+	// The reason is that gov.ReadGovernance() depends on the caches, and that
+	// gov.WriteGovernance() sets idxCache accordingly, whereas db.WriteGovernance don't,
```
Comment (yoomee1313) [inline governance/mixed_test.go:81]:
```suggestion
	// gov.WriteGovernance() sets idxCache accordingly, whereas db.WriteGovernance don't
```

**Review by yoomee1313 — COMMENTED** [review]
looks good to me except miner typos

**Review by yoomee1313 — APPROVED** [review]

**Review by ian0371 — APPROVED** [review]

**Pull request merged**